pulp_ansible/app/tasks/collections.py

```diff
@@ -460,8 +473,52 @@ def _collection_list_downloader(self, api_version, collection_endpoint, page_num
             collection_list_url = f"{collection_endpoint}?limit={page_size}&offset={offset}"
         return self.remote.get_downloader(url=collection_list_url)
 
+    async def _find_all_unpaginated_collections(self):
+        tasks = []
+        loop = asyncio.get_event_loop()
+
+        root_endpoint, api_version = await self._get_root_api(self.remote.url)
+        collection_endpoint = f"{root_endpoint}/metadata/collections/"
+        if api_version > 2:
+            downloader = self.remote.get_downloader(
+                url=collection_endpoint, silence_errors_for_response_status_codes={404}
+            )
+            try:
+                metadata = parse_metadata(await downloader.run())
+            except FileNotFoundError:
+                pass
+            else:
+                for collection_data in metadata:
+                    if collection_data["deprecated"]:
+                        self.deprecations |= Q(
+                            namespace=collection_data["namespace"], name=collection_data["name"]
+                        )
+                collection_version_endpoint = f"{root_endpoint}/metadata/collection_versions/"
+                downloader = self.remote.get_downloader(url=collection_version_endpoint)
+                metadata = parse_metadata(await downloader.run())
+                for collection_version_metadata in metadata:
+                    collection_version_url = urljoin(
+                        self.remote.url, f"{collection_version_metadata['href']}"
+                    )
+                    tasks.append(
+                        loop.create_task(
+                            self._add_collection_version(
+                                api_version, collection_version_url, collection_version_metadata
+                            )
+                        )
+                    )
+
+                await asyncio.gather(*tasks)
+                return True
+
+        return False
+
     async def _find_all_collections(self):
-        collection_endpoint, api_version = await self._get_collection_api(self.remote.url)
+        found = await self._find_all_unpaginated_collections()
```

Comment:
The issue with calling this new workflow from here is that it doesn't trigger when using a requirements file. I think the new metadata format needs to be used when using a requirements file too whenever its available.

I was testing this with `pytest -v -r sx --color=yes --pyargs pulp_ansible.tests.functional.api.collection.v3.test_sync::SyncCollectionsFromPulpServerTestCase::test_sync_collections_from_pulp`

Reply:
I intentionally added it here, why would I want a huge blob of collections when I have a little requirements file?

Reply:
I think of a few reasons.
* We need to cut down on the number of requests becasue it triggers rate limiting
* If we configure `rate_limit` making the 3*N calls for the content in the requirements file will likely take much longer than requesting 1 larger response once
* Consider requirements files with many collections in them, like hundreds.

Reply:
I agree with all your reasons, but take our tests,
we generally do:
```
"collections:\n  - testing.k8s_demo_collection"
```
Let's say we have a repo with 30000 collections and this one is in the middle, I don't think it would be good to loop many collections to get just a few.

Maybe we can adopt different strategies depending on `len(self.collection_info)`

Reply:
I can see your point, but consider the actual execution times here. This isn't exactly an apples for apples because there is json processing time and more comparisons, etc, but here's a little script I wrote that loops over 30,000 tuples checking each one:

```
def test():
    """Test function"""
    for item in data:
        if item[1] == 'yes':
            var = 1

data = [('a', 'yes') * 30000]

if __name__ == '__main__':
    import timeit
    print(timeit.timeit("test()", setup="from __main__ import test, data"))
```

When I run this it takes `0.11673803627490997` seconds.

I was then pinging galaxy to show it's network round trip time by comparison:

```
[bmbouter@localhost tmp]$ ping galaxy.ansible.com
PING galaxy.ansible.com (104.26.0.234) 56(84) bytes of data.
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=1 ttl=55 time=25.0 ms
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=2 ttl=55 time=31.6 ms
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=3 ttl=55 time=36.9 ms
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=4 ttl=55 time=30.0 ms
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=5 ttl=55 time=27.7 ms
64 bytes from 104.26.0.234 (104.26.0.234): icmp_seq=6 ttl=55 time=27.6 ms
```

Say we say it takes 27ms which is .027. seconds. So if we even make 4 requests looping over 30K items would still be faster.

This isn't a great experiement for a lot of reasons but my point is that fetching a single blob of data that is large, but only having to go to the server once is very efficient.